llvm/lib/Transforms/Utils/LoopMetadata.cpp

```diff
@@ -0,0 +1,477 @@
+//===-- LoopMetadata.cpp - Convert Loop Metadata to assumes --*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopMetadata.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+
+#define DEBUG_TYPE "loop-metadata"
+
+using namespace llvm;
+
+bool isRotatable(const Loop *L);
+///  Check if the minimum value fits into the given type
+bool fitstype(unsigned MinValue, const Type *T);
+
+/// extract the SCEVAddExpr that is in the operands of the Instr
+const SCEVAddExpr *getAddExpr(const Instruction *Instr, ScalarEvolution *SE);
+
+/// get SCEVTruncExpr from the SCEV S
+const SCEVTruncateExpr *getSCEVTruncate(const SCEV *S);
+
+/// extract the step size by which the loop IV changes.
+int getSCEVStepSize(const SCEV *S);
+
+/// based on the SCEVTruncateExpr extract the Loop Start Value
+Value *getSCEVStart(const SCEVTruncateExpr *S, const BasicBlock *LoopPreHeader,
+                    const ScalarEvolution *SE);
+
+PreservedAnalyses LoopMetadata::run(Loop &L, LoopAnalysisManager &AM,
+                                    LoopStandardAnalysisResults &AR,
+                                    LPMUpdater &U) {
+  SE = &AR.SE;
+  AC = &AR.AC;
+  DT = &AR.DT;
+
+  assignLoopMetadata(L);
+
+  return PreservedAnalyses::all();
+}
+
+bool LoopMetadata::assignLoopMetadata(Loop &L) {
+  this->L = &L;
+  Context = &L.getHeader()->getParent()->getContext();
+
+  std::optional<int> MinIterCount = getMinTripCount(&L);
+
+  // dump loop summary
+  LLVM_DEBUG(if (L.getLoopPreheader()) {
+    dbgs() << "Preheader:" << L.getLoopPreheader()->getName() << "\n";
+  });
+  LLVM_DEBUG(dbgs() << "Header:" << L.getHeader()->getName() << "\n");
+
+  if (MinIterCount.has_value() && MinIterCount.value() > 0) {
+    this->MinIterCount = MinIterCount.value();
+
+    if (!isRotatable(this->L)) {
+      LLVM_DEBUG(dbgs() << "Processing Loop Metadata: "
+                        << L.getHeader()->getParent()->getName() << " "
+                        << L.getName() << " (" << this->MinIterCount
+                        << ")\nAborting Metadata due to not rotatable!\n");
+      return false;
+    }
+
+    LLVM_DEBUG(L.getHeader()->getParent()->dump(););
+    addAssumeToLoopHeader();
```

Comment:
I see that `addAssumeToLoopHeader` is returning early due to several reasons. Should we get this early returns as a boolean type and then `return addAssumeToLoopHeader();` ?